src/coreclr/jit/gentree.h

```diff
@@ -540,6 +540,8 @@ enum GenTreeFlags : unsigned int
 
     GTF_DIV_MOD_NO_OVERFLOW     = 0x40000000, // GT_DIV, GT_MOD -- Div or mod definitely does not overflow.
 
+    GTF_UMOD_UINT16_OPERANDS    = 0x80000000, // UMOD -- Both operands to a mod are in uint16 range.
```

Comment:
This flag needs to be checked by `GenTree::Compare` (reason N+1 that flags changing IR semantics is unfortunate IR design)

Reply:
I don't think we can just avoid it (not here, but in general). But perhaps we can introduce a separate Enum with changing-semantics flags or something like that

Reply:
I don't see why this would need to be taken into account in `GenTree::Compare`. This flag, similar to `GTF_IND_NONFAULTING`, only records an optimization fact.

Reply:
Ideally operations with different semantics should have different `gtOper`, like e.g. `BSWAP` and `BSWAP16` do. That might not be as convenient to do in some cases, but ideally we should optimize the process of adding new types of nodes to be as painless as possible.

Reply:
> I don't see why this would need to be taken into account in GenTree::Compare. This flag, similar to GTF_IND_NONFAULTING, only records an optimization fact.

For example head-merging a 16-bit division with a non-16-bit division would not be a legal transformation.

Reply:
I don't disagree that a better way to handle this would be to introduce a `GT_UMOD16` that would perform true 16-bit division, ignoring the upper bits of the operands.

But in this flags model, there is no true 16 division, only 32 bit division "for which we know the operands fit into a certain range". It would be invalid IR for a node tagged with the flag to appear where the dynamic range of the operands does not actually fit. Hence, `Compare` ought not to care.

What needs to care is all the places that manipulated `UMOD` nodes.

Reply:
> It would be invalid IR for a node tagged with the flag to appear where the dynamic range of the operands does not actually fit.

Yes, and this can happen in head-merging if it merges a 16-bit division and a non-16-bit division and moves them across a guard on the divisor range. I view that as changing the semantics of the node (the domain of the operands is, yet the value computed is different for a divisor outside [0..0xffff]).

Reply:
I do not think we can model this as an optimization fact. It requires us to set `GTF_ORDER_SIDEEFF` on both this node and its guarding condition IR node, similar to other unmodelled control-flow dependencies we have.

Reply:
Hmm, I see what you are saying (that this flag _should_ mean "a kind of true 16 bit division"). Perhaps the flag could use a better name then.